planner/core/plan.go

```diff
@@ -210,17 +213,47 @@ func (p *baseLogicalPlan) storeTask(prop *property.PhysicalProperty, task task)
 	p.taskMap[string(key)] = task
 }
 
-// BuildKeyInfo implements LogicalPlan BuildKeyInfo interface.
-func (p *baseLogicalPlan) BuildKeyInfo() {
-	for _, child := range p.children {
-		child.BuildKeyInfo()
-	}
-	switch p.self.(type) {
-	case *LogicalLock, *LogicalLimit, *LogicalSort, *LogicalSelection, *LogicalApply, *LogicalProjection:
-		p.maxOneRow = p.children[0].MaxOneRow()
+// HasMaxOneRow returns if the LogicalPlan will output at most one row.
+func HasMaxOneRow(p LogicalPlan, childMaxOneRow []bool) bool {
+	// The reason why we use the check `len(childMaxOneRow) == ?` below
+	// is that, this function is used both in planner/core and planner/cascades.
+	// In cascades planner, LogicalPlan may have no `children`.
+	switch x := p.(type) {
+	case *LogicalLock, *LogicalLimit, *LogicalSort, *LogicalSelection,
+		*LogicalApply, *LogicalProjection, *LogicalWindow, *LogicalAggregation:
+		if len(childMaxOneRow) == 1 {
+			return childMaxOneRow[0]
+		}
 	case *LogicalMaxOneRow:
-		p.maxOneRow = true
+		return true
+	case *LogicalJoin:
+		switch x.JoinType {
+		case SemiJoin, AntiSemiJoin, LeftOuterSemiJoin, AntiLeftOuterSemiJoin:
+			if len(childMaxOneRow) == 1 {
```

Comment:
Why do we use `len(childMaxOneRow) == 1`? normally `len(childMaxOneRow` would be 2 when coming to here?